manifests/scheduled_repo.pp

```diff
@@ -0,0 +1,38 @@
+# @summary A wrapper that creates a pulp_Xrepo resource with and an associated pulp_schedule resource
+#
+# @example Basic usage with default `rpm` `repo_type` and 'random' daily schedule
+#   pulp::scheduled_repo { 'centos-7':
+#     repo_config => {
+#       'display_name' => 'CentOS 7 Base Repo',
+#       'feed'         => 'https://www.mirrorservice.org/sites/mirror.centos.org/7/os/x86_64',
+#     },
+#   }
+#
+# @param repo_type
+#   Determines whether a `pulp_rpmrepo`, `pulp_puppetrepo` or `pulp_isorepo` resource is created.
+# @param repo_config
+#   Defines a hash of parameters to pass to the `pulp_Xrepo` resource.
+# @param schedule
+#   An iso8601 schedule string or the special value 'daily'. Defaults to 'daily' which creates a `pulp_schedule` with a pseudo-random daily `schedule_time`.
+define pulp::scheduled_repo
+(
+  Enum['rpm','puppet','iso'] $repo_type = 'rpm',
+  Hash $repo_config = {},
+  Variant[Enum['daily'],String[1]] $schedule = 'daily',
+)
+{
+  $schedule_time = $schedule ? {
+    'daily' => pulp::daily_schedule($name),
+    default => $schedule,
+  }
+
+  case $repo_type { # lint:ignore:case_without_default
```

Comment:
a `default` would be unreachable since `$repo_type` is an `Enum`

Reply:
An alternative that might work:
```puppet
"pulp_${repo_type}repo" { $name:
  * => $repo_config,
}
```

Reply:
ooh.  I didn't think this would work.  Now that I've written some tests, it's easy to find out.

Reply:
`Error while evaluating a Resource Statement, Evaluation Error: Illegal Resource Type expression, expected result to be a type name, or untitled Resource, got String`

Reply:
I guess `create_resource` could be used and I think here it's actually cleaner

Reply:
Really? I thought that was a left over from the puppet 3 days and considered evil.

Reply:
Evil to some people anyway...
https://puppet.com/docs/puppet/3.8/future_lang_resources_advanced.html#implementing-the-createresources-function
https://www.craigdunn.org/2016/04/designing-modules-in-a-puppet-4-world

Reply:
But in those cases the resource type is not variable :)

Reply:
This looks weird, but should work. (eyeball compiled only)

```
Resource["pulp_${repo_type}repo"] { $name:
  * => $repo_config,
}
```

Reply:
I should clarify that I'm not yet convinced that I like this pattern yet. It's... not exactly intuitively readable unless you use mental gymnastics like this:

> index into all *resources* to find the type  "pulp_${repo_type}repo"

Reply:
I wonder this would be easier to read:
```puppet
$resource_type = Resource["pulp_${repo_type}repo"]
$resource_type { $name:
  * => $repo_config,
}
```

Feels a bit redundant, but my mind has trouble with the uppercase `Resource`. Makes me think it refers to an existing instance rather than defining a new one.

Reply:
Easier to read, but no longer valid. :)
`Evaluation Error: Illegal Resource Type expression, expected result to be a type name, or untitled Resource, got Type[Pulp_rpmrepo]`

@binford2k's example does work though.